File: thanh123456_AcceptedCodeVNSPOJ/RIDDLE.pas
const
     fin='';fon='';
     maxn=trunc(1e6);
type
     mang1=array [0..maxn] of longint;
     mang2=array [0..maxn] of int64;
var
     n,k,test:longint;
     a,b:mang1;
     s:mang2;
procedure nhap;
     begin
          assign(input,fin);reset(input);
          assign(output,fon);rewrite(output);
          readln(tesT);
     end;
procedure nhapdulieu;
     var
          i:longint;
     begin
          readln(n,k);
          for i:=1 to n do read(b[i]);
     end;
procedure swap(VAr x,y:longint);
     var
          tam:longint;
     begin
          tam:=x; x:=y; y:=tam;
     end;
procedure qsort(l,r:longint);
     var
          i,j,x,m:longint;
     begin
          i:=l; j:=r;
          m:=random(r-l)+l+1;
          x:=a[m];
          repeat
               while a[i]<x do inc(i);
               while a[j]>X do dec(j);
               if i<=j then
                    begin
                         swap(a[i],a[j]);
                         inc(i); dec(j);
                    end;
          until i>j;
          if l<j then qsort(l,j);
          if i<r then qsort(i,r);
     end;
function Check(m:longint):boolean;
     var
          i:longint;
     begin
          for i:=1 to m do a[i]:=b[i];
          qsort(1,m);
          for i:=1 to m do
               begin
                    if s[i-1]+1<int64(a[i]) then exit(false);
                    s[i]:=s[i-1]+int64(a[i]);
               end;
          if s[m]<k then exit(false);
          exit(true);
     end;
procedure solve;
     var
          l,r,mid,res:longint;
     begin
          l:=1;
          res:=0;
          r:=n;
          repeat
               mid:=(l+r) div 2;
               if Check(mid)=false then
                    l:=mid+1
               else
                    begin
                         res:=mid;
                         r:=mid-1;
                    end;
          until l>r;
          if res=0 then writeln(-1)
          else writeln(res);
     end;
procedure xuat;
     var
          i:longint;
     begin
          for i:=1 to test do
               begin
                    nhapdulieu;
                    solve;
               end;
          close(input); close(output);
     end;
begin
     nhap;
     xuat;
end.

File: thanh123456_AcceptedCodeVNSPOJ/RIDDLE.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 5e5;
int n,test,k,a[N],b[N];
void nhap()
{
    scanf("%d",&test);
}
void doc()
{
    scanf("%d%d",&n,&k);
    memset(a,0,sizeof(a));
    for (int i=1; i<=n; i++) {scanf("%d",&b[i]);}
    //sort(a+1,a+1+n);
}
bool check(int m)
{
	for (int i=1; i<=m; i++) {a[i] = b[i];}
	sort(a+1,a+1+m);
    long long s = 0;
    for (int i=1; i<=m; i++)
    {
        if (s+1 < a[i]) {return false;}
        s+=(long long)a[i];
    }
    if (s<k) {return false;}
    return true;
}
void giai()
{
    int l = 1;
    int r = n;
    int res = 0;
    while (l<=r)
    {
        int mid = (l+r) /  2;
        if (check(mid)) {res=mid; r=mid-1;}
        else {l=mid+1;}
    }
    printf("%d\n",res ? res : -1);
}
void solve()
{
    for (int i=1; i<=test; i++)
    {
        doc();
        giai();
    }
}
int main()
{
    nhap();
    solve();
}
